chautnguyen4392/yacoin
# Add headers-first logic and improve P2P logic (#100)

* Add debug log

* Replace pindexBest, pindexGenesisBlock, nBestHeight with chainActive

* Replace bnBestInvalidTrust with pindexBestInvalid

* Clang-format txdb-leveldb.cpp

* Add the missing logic: bnChainTrust is calculated at the startup

* Add setBlockIndexValid, BlockStatus enum, and new flag nStatus for
CBlockIndex

* Use setBlockIndexValid and new flag nStatus for CBlockIndex

* Add ConnectBestBlock()

* Refactor SetBestChain and Reorganize logic

* Add CValidationState

* Move only: extract WriteChainState and UpdatedTip from SetBestChain.

* Prepare block connection logic for headers-first

* Remove redundant code

* Reindex for old database

* clang format network logic

* Make sure we always have a node to do IBD from

This introduces the concept of the 'sync node', which is the one we
asked for missing blocks. In case the sync node goes away, a new one
will be selected.

For now, the heuristic is very simple, but it can easily be extended
later to add better policies.

* Add main-specific node state & move ban score

* Per-peer block download tracking and stalled download detection.

* Split off CBlockHeader from CBlock

* Split up CheckBlock in a block and header version

* Split AcceptBlockHeader from AcceptBlock

* Allow SendMessages to run partially without cs_main

* Get rid of the static chainMostWork (optimization)

* Allow ActivateBestChain to release its lock on cs_main

* Move all post-chaintip-change notifications to ActivateBestChain

* Push cs_mains down in ProcessBlock

* Add missing LOCK(cs_main)

* fixup! Move all post-chaintip-change notifications to ActivateBestChain

* Track peers' available blocks

* Add a skiplist to the CBlockIndex structure.

* Replace FindBlockByHeight with chainActive height

* Replace CBlockLocator with chainActive.GetLocator

* Correct "detect stalled peers" logic

* Headers-first synchronization

* RPC additions after headers-first

* Remove checks to prevent "fill up memory by spamming with bogus blocks", as we always know all parent headers

* Better logging

* Improve performance of ActivateBestChainStep

Improve the performance of ActivateBestBlockStep by using the skiplist to only discover
a few potential blocks to connect at a time, instead of all blocks forever - as we likely bail
out after connecting a single one anyway.

* Fix large reorgs

* Avoid calculating hash many times at initial-block-sync

* Remove redundant debug log

* Fix CChain::GetLocator

* Only keep setBlockIndexValid entries that are possible improvements

* Not return error when Rejected by stake modifier checkpoint

* Fix issue which update wrong block index in ReceivedBlockTransactions

* Avoid sending "getheaders" when receive "inv" during IBD

* fixup! Avoid calculating hash many times at initial-block-sync

* fixup! fixup! Avoid calculating hash many times at initial-block-sync

* fixup! Not return error when Rejected by stake modifier checkpoint

* Add extra checkpoints

* Various improvement for initial-block-sync

* Timeout for headers sync

* Not download block from initial-header-sync node, unless best header close to today

* Trigger sending "getblocks" to other peers when enough headers

Trigger sending "getblocks" to other peers when
1) bestHeaderHeight > bestBlockHeight + HEADER_BLOCK_DIFFERENCES_TRIGGER_GETDATA (default = 400000)
2) synced_headers < bestHeaderHeight

* Reindex for old database (part 2)

* Fix issue which failed to accept hardfork block

* Correct the logic which recalculate block reward and minEase when starting node

* fixup! Correct the logic which recalculate block reward and minEase when starting node

* Lower the DoS score when incorrect Pow to not ban the node which have weaker-chain

* Remove redundant release time

* Add debug log when FinalizeNode

* Timeout for block sync

* Allow user to configure values relating to init-sync

* Avoid spamming getblocks, just send getblocks every 1 minute

* Add missing CheckBlockHeader in AcceptBlockHeader

* Avoid syncing headers from the peer which doesn't have latest blockchain

* Fix log

* Change getblockchaininfo to gettimechaininfo
